### DISCUSSION TITLE
Semantics of using "job.needs" and "job.if" together

### DISCUSSION THREAD

**Discussion by hiranya911** [community/community#25286] [Actions]
<p>The documentation at <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idneeds" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idneeds</a> states:</p>
<pre><code>If a job fails, all jobs that need it are skipped unless the jobs use a conditional statement
that causes the job to continue.
</code></pre>
<p>Now suppose I have a job configured as follows:</p>
<pre><code>job2:
  needs: job1
  if: true
</code></pre>
<p>Will job2 always run regardless of the outcome of job1? Will it at least wait until job1 finishes?</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/hiranya911">@hiranya911</a> ,</p>
<p>Job2 will be executed until job1 finished. As doc mentioned, if job1 failed, job2 will be skipped.</p>
<p>The exception is, if you set ‘if: always()’ in job2, job2 will be executed even job1 failed. Please refer to the official <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#always" rel="nofollow noopener">doc</a> for more details.</p>
<pre><code>job2:
    runs-on: [ubuntu-latest]
    if: always()
    needs: job1
</code></pre>
<p><img src="https://user-images.githubusercontent.com/52772761/181090078-07f5e87c-fdfd-4bf5-89b1-951bfb5972f9.png" alt="always.png" data-base62-sha1="3N0X64RYpJau55oPezb2vd510FN" width="578" height="324"></p>

**Comment by hiranya911** [comment]
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a> </p>
<p>It’s great to know that <code>always()</code> works in the context of chained jobs too. I also tried the following, as a further experiment, and thought I’d share my results:</p>
<pre><code>jobs:
  job1:
    runs-on: ubuntu-latest
    steps:
    - run: |
        echo Job 1
        exit 1

  job2:
    runs-on: ubuntu-latest
    needs: job1
    if: github.event.client_payload.job2
    steps:
    - run: echo Job 2
</code></pre>
<p>Here job2 never runs (even if the client_payload.job2 is set to true), since the job1 is guaranteed to fail. This is exactly the semantics I needed for my workflow.</p>

**Comment by maxisam** [comment]
<p>you have to use always() it is a special flag</p>
